controller/controller_client.go

```diff
@@ -23,266 +20,234 @@ type doneChan chan struct{}
-func (c *clientHandler) shutdown(shipperUnit *client.Unit) {
-	swapped := atomic.CompareAndSwapUint32(&c.shipperIsStopping, 0, 1)
-	if !swapped {
-		return
-	}
-	_ = shipperUnit.UpdateState(client.UnitStateStopping, "shutting down shipper output", nil)
-	c.stopShipper()
-	// The server has successfully shut down
-	// In theory we want this to block, as we should wait for queues & outputs to empty.
-	c.shutdownComplete.Wait()
-	c.log.Debugf("Shutdown complete, sending STOPPPED")
-	_ = shipperUnit.UpdateState(client.UnitStateStopped, "Stopped shipper output", nil)
-}
-
-// runController is the main runloop for the shipper itself, and managed communication with the agent.
-func runController(ctx context.Context, agentClient client.V2) error {
-	log := logp.L()
-
-	err := agentClient.Start(ctx)
-	if err != nil {
-		return fmt.Errorf("error starting connection to client")
-	}
-
-	log.Debugf("Starting error reporter")
-	go reportErrors(ctx, agentClient)
-
-	log.Debugf("Started client, waiting")
-
-	handler := newClientHandler()
-
-	// receive the units
-	for {
-		select {
-		case <-ctx.Done():
-			handler.log.Debugf("Got context done")
-			shipperUnit := handler.getUnit(handler.shipperOutputID)
-			handler.shutdown(shipperUnit)
-			// If we get context done, just hard-stop
-			return nil
-
-		case change := <-agentClient.UnitChanges():
-
-			switch change.Type {
-			case client.UnitChangedAdded: // The agent is starting the shipper, or we added a new processor
-				go handler.handleUnitAdded(change.Unit)
-			case client.UnitChangedModified: // config for a unit has changed
-				go handler.handleUnitUpdated(change.Unit)
-			case client.UnitChangedRemoved: // a unit has been stopoped and can now be removed.
-				handler.deleteUnit(change.Unit)
-				// for now, consider a remove of the shipper unit to be our sign to shut down.
-				// Take care, as we won't get this until we've sent a STOPPED to the agent
-				// TODO: we should have a timeout so we can shutdown without getting a REMOVED event
-				if change.Unit.Type() == client.UnitTypeOutput {
-					handler.log.Debugf("shipper unit removed, ending.")
-					return nil
-				}
-
-			}
-		}
+		c.units.UpdateUnit(unit)
 	}
+	return needsRestart
 }
 
-// I am not net sure how this should work or what it should do, but we need to read from that error channel
-func reportErrors(ctx context.Context, agentClient client.V2) {
-	log := logp.L()
-	for {
-		select {
-		case <-ctx.Done():
-			return
-		case err := <-agentClient.Errors():
-			log.Errorf("Got error from controller: %s", err)
-		}
+// handle the UnitChangedRemoved event from the V2 API
+// returns true if the shipper component needs to be bounced.
+func (c *clientHandler) handleUnitRemoved(unit *client.Unit) bool {
+	state, logLvl, _ := unit.Expected()
+	c.log.Debugf("Got unit removed for ID %s (%s/%s)", unit.ID(), state.String(), logLvl.String())
+	unitType := unit.Type()
+	if unitType == client.UnitTypeOutput {
```

Comment:
```suggestion
	if unitType == client.UnitTypeInput {
```

Looks like this might be reversed base on the body of the if condition calling `DeleteInput`.